File: src/main/java/sql/createdf/RDD2DataFrameByProgrammatically.java
package main.java.sql.createdf;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;


public class RDD2DataFrameByProgrammatically {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setMaster("local").setAppName("RDD2DataFrameByProgrammatically");
		JavaSparkContext sc = new JavaSparkContext(conf);
		SQLContext sqlcontext = new SQLContext(sc);
		/**
		 * 在RDD的基础上创建类型为Row的RDD
		 */
		JavaRDD<String> lines = sc.textFile("Peoples.txt");
		JavaRDD<Row> rowRDD = lines.map(new Function<String, Row>() {
			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public Row call(String line) throws Exception {
				String[] split = line.split(",");
				return RowFactory.create(Integer.valueOf(split[0]),split[1],Integer.valueOf(split[2]));
			}
		});
		
		/**
		 * 动态构造DataFrame的元数据，一般而言，有多少列以及每列的具体类型可能来自于Json，也可能来自于DB
		 */
		ArrayList<StructField> structFields = new ArrayList<StructField>();
		structFields.add(DataTypes.createStructField("id", DataTypes.IntegerType, true));
		structFields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
		structFields.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));
		//构建StructType，用于最后DataFrame元数据的描述
		StructType schema = DataTypes.createStructType(structFields);
		
		/**
		 * 基于已有的MetaData以及RDD<Row> 来构造DataFrame
		 */
		DataFrame df = sqlcontext.createDataFrame(rowRDD, schema);
		
		/**
		 *注册成为临时表以供后续的SQL操作查询
		 */
		df.registerTempTable("persons");
		
		/**
		 * 进行数据的多维度分析
		 */
		DataFrame result = sqlcontext.sql("select * from persons where age > 7");
		result.show();

		/**
		 * 对结果进行处理，包括由DataFrame转换成为RDD<Row>
		 */
	     List<Row> listRow = result.javaRDD().collect();
		 for (Row row : listRow) {
			System.out.println(row);
		}
	}
}

File: src/main/scala/sql/createdf/RDD2DataFrameByProgrammatically.scala
package main.scala.sql.createdf

import org.apache.spark.sql.SQLContext
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql.types.StructField
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.types.StringType
import org.apache.spark.sql.Row
import org.apache.spark.sql.types.IntegerType
import java.util.Arrays.ArrayList
import java.util.ArrayList

object RDD2DataFrameByProgrammatically {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf() //创建sparkConf对象
    conf.setAppName("My First Spark App") //设置应用程序的名称，在程序运行的监控页面可以看到名称
    conf.setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)
    val people = sc.textFile("scores.txt")
    
    val schemaString = "clazz:String score:Integer"
    //如果schema中制定了除String以外别的类型   在构建rowRDD的时候要注意指定类型     例如： p(2).toInt 
    val rowRDD = people.map(_.split("\t")).map(p => Row(p(0), p(1).toInt))
    
    val schema =
      StructType(schemaString.split(" ").map(fieldName => StructField(fieldName.split(":")(0), if (fieldName.split(":")(1).equals("String")) StringType else IntegerType, true)))
//    val structFields = Array(StructField("clazz",StringType,true),StructField("score",IntegerType))
//    val schema = StructType(structFields)
    
    
    
    //  val arr = Array(StructField("name",StringType,true),StructField("age",IntegerType,true))
    //  val schema = StructType.apply(arr)
    val peopleDataFrame = sqlContext.createDataFrame(rowRDD, schema)
    peopleDataFrame.printSchema()
    peopleDataFrame.show()
   /* peopleDataFrame.registerTempTable("clazzScore")
    val results = sqlContext.sql("SELECT score,clazz FROM clazzScore")
    //  results.map(t => "age: " + t(0)).collect().foreach(println)
    results.map(t => "clazz: " + t.getAs[String]("clazz")+"\tscore:"+t.getAs[Integer]("score")).foreach(println)*/
    
  }
}












